File: codes/poj3041.cpp
#include<cstdlib>
#include<cstdio>
#include<vector>
#include<cstring>
using namespace std;

vector<int> g[503];
int p[503];
bool vis[503];
int n;

bool find(int u)
{
  for (int i=0;i<g[u].size();i++)
  {
    if (!vis[g[u][i]])
    {
      vis[g[u][i]]=true;
      if (!p[g[u][i]]||find(p[g[u][i]]))
      {
        p[g[u][i]]=u;
        return true;
      }
    }
  }
  return false;
}

int main()
{
  int m,j,i;
  //freopen("3041.in","r",stdin);
  //freopen("3041.out","w",stdout);
  scanf("%d%d",&n,&m);
  for (i=0;i<503;i++) g[i].clear();
  memset(p,0,sizeof(p));
  for (i=1;i<=m;i++)
  {
    int u,v;
    scanf("%d%d",&u,&v);
    g[v].push_back(u);
  }  
  int ans=0;
  for (i=1;i<=n;i++)
  {
    memset(vis,0,sizeof(vis));
    if (find(i)) ans++;  
  }
  printf("%d\n",ans);
  return 0;
}

File: PKU-Online-Judge/poj3041.pas
var
  map:array[0..501,0..501]of boolean;
  vis:array[0..501]of boolean;
  p:array[0..501]of longint;
  i,j,k,m,n,s,t,ans:longint;
  function find(i:longint):boolean;
    var
      j:longint;
    begin
      for j:=1 to n do
        if (not vis[j])and(map[i,j]) then
          begin
            vis[j]:=true;
            if (p[j]=0)or(find(p[j])) then
              begin
                p[j]:=i;
                exit(true);
              end;
          end;
      exit(false);
    end;
begin
  fillchar(map,sizeof(map),false);
  fillchar(p,sizeof(p),0);
  readln(n,m);
  for i:=1 to m do
    begin
      readln(s,t);
      map[s,t]:=true;
    end;
  ans:=0;
  for i:=1 to n do
    begin
      fillchar(vis,sizeof(vis),false);
      if find(i) then
        inc(ans);
    end;
  writeln(ans);
end.
